File: algorithms/0-9/4Sum/4Sum.js
var fourSum = function(nums, target) {
    nums.sort((a, b) => a - b);
    const res = [];
    
    for(let i=0; i<nums.length-3; i++) {
        for(let j=i+1; j<nums.length-2; j++) {
            let min = j + 1;
            let max = nums.length - 1;
            while(min < max) {
                const sum = nums[i] + nums[j] + nums[min] + nums[max];
                if(sum === target) {
                    res.push([nums[min], nums[max], nums[i], nums[j]]);
                    while(nums[min] === nums[min+1]) min++;
                    while(nums[max] === nums[max-1]) max--;
                    min++;
                    max--;
                }
                
                else if(sum > target) 
                    max--;
                else
                    min++
            }
            while(nums[j] === nums[j+1]) j++;
        }
        while(nums[i] === nums[i+1]) i++;
    }
    
    return res;
};

File: scripts/algorithms/0-9/4Sum/4Sum.java
// Runtime: 17 ms (Top 82.83%) | Memory: 42.5 MB (Top 98.57%)
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> llans = new LinkedList<>();
        if(nums == null || nums.length <= 2){
            return llans;
        }
        for(int i=0;i<nums.length-3;i++){
            for(int j=i+1;j<nums.length-2;j++){
                int l = j+1;
                int r = nums.length-1;
                long ressum = target - nums[i];
                ressum = ressum - nums[j];
                while(l<r){
                    if(nums[l]+nums[r] == ressum){
                        List<Integer> ll = new LinkedList<>();
                        ll.add(nums[i]);
                        ll.add(nums[j]);
                        ll.add(nums[l]);
                        ll.add(nums[r]);
                        llans.add(ll);

                        while( l<r && nums[l]==nums[l+1]){l++;}
                        while( l<r && nums[r]==nums[r-1]){r--;}
                        l++;
                        r--;
                    }
                    else if(nums[l]+nums[r] < ressum){
                        l++;
                    }
                    else{
                        r--;
                    }
                }
                while( j<nums.length-1 && nums[j]==nums[j+1]){
                    j++;}

            }
            while( i<nums.length-1 && nums[i]==nums[i+1]){
                i++;}

        }
        return llans;
    }
}